sycl/source/detail/pi.cpp

```diff
@@ -129,39 +129,99 @@ std::string memFlagsToString(pi_mem_flags Flags) {
+  const char *Env = std::getenv(SYCL_PI_TRACE);
+  m_Data = (Env ? std::atoi(Env) : 0);
+}
+
+static const std::map<std::string, Backend> SyclBeMap{
+    {"PI_OTHER", SYCL_BE_PI_OTHER},
+    {"PI_CUDA", SYCL_BE_PI_CUDA},
+    {"PI_OPENCL", SYCL_BE_PI_OPENCL}};
+
+template <> Config<Backend, SYCL_BE>::Config() {
+  const char *BE = std::getenv(SYCL_BE);
+  if (BE) {
+    auto It = SyclBeMap.find(BE);
+    if (It == SyclBeMap.end())
+      pi::die("Invalid SYCL_BE. Valid values are PI_OPENCL/PI_CUDA");
+    m_Data = It->second;
+  } else {
+    // Default preference is to OpenCL BE.
+    m_Data = SYCL_BE_PI_OPENCL;
+  }
 }
 
-// Check for manually selected BE at run-time.
-bool useBackend(Backend TheBackend) {
-  return TheBackend == getBackend();
+// SYCL_INTEROP_BE is a way to specify the interoperability plugin.
+template <> Config<Backend, SYCL_INTEROP_BE>::Config() {
+  const char *BE = std::getenv(SYCL_INTEROP_BE);
+  if (BE) {
+    auto It = SyclBeMap.find(BE);
+    if (It == SyclBeMap.end())
+      pi::die("Invalid SYCL_INTEROP_BE. "
+              "Valid values are PI_OPENCL/PI_CUDA");
+    m_Data = It->second;
+  } else {
+    // Default preference is to OpenCL.
+    m_Data = SYCL_BE_PI_OPENCL;
+  }
 }
 
+// Helper interface to not expose "pi::Config" outside of pi.cpp
+Backend getPreferredBE() { return Config<Backend, SYCL_BE>::get(); }
+
 // GlobalPlugin is a global Plugin used with Interoperability constructors that
 // use OpenCL objects to construct SYCL class objects.
 std::shared_ptr<plugin> GlobalPlugin;
 
 // Find the plugin at the appropriate location and return the location.
-// TODO: Change the function appropriately when there are multiple plugins.
-bool findPlugins(vector_class<std::string> &PluginNames) {
+bool findPlugins(vector_class<std::pair<std::string, Backend>> &PluginNames) {
   // TODO: Based on final design discussions, change the location where the
   // plugin must be searched; how to identify the plugins etc. Currently the
   // search is done for libpi_opencl.so/pi_opencl.dll file in LD_LIBRARY_PATH
   // env only.
-  PluginNames.push_back(OPENCL_PLUGIN_NAME);
-  PluginNames.push_back(CUDA_PLUGIN_NAME);
+  //
+  // NOTE: the order of the plugins below *does* matter. Some custom platform/
+  // device selection codes, which do not distinguish different backends would
+  // (per current logic in device_selector::select_device()) select first
+  // device that fits. SYCL spec says: "If more than one device receives
+  // the high score then one of those tied devices will be returned, but
+  // which of the devices from the tied set is to be returned is not defined".
+  //
+  // By having OpenCL plugin seen first we give a slight and subtle preference
+  // to OpenCL devices.
+  //
+  PluginNames.push_back(std::make_pair<std::string, Backend>(
+      OPENCL_PLUGIN_NAME, SYCL_BE_PI_OPENCL));
+  PluginNames.push_back(
+      std::make_pair<std::string, Backend>(CUDA_PLUGIN_NAME, SYCL_BE_PI_CUDA));
```

Comment:
Looks like this will force the OpenCL NVIDIA platform to come before the CUDA backend...

Reply:
That's correct, by default it would be OpenCL. The explicit SYCL_BE=PI_CUDA would change that to CUDA.

Reply:
Let's please move this preference inside the device_selector::select_device(). It is too subtle to have it like here. In select_device() you'd just choose the device of the preferred BE if multiple BE get the same score.

Reply:
Done